File: Makwa/c/deleggen.c
/*
 * -----------------------------------------------------------------------
 * (c) Thomas Pornin 2014. This software is provided 'as-is', without
 * any express or implied warranty. In no event will the author be held
 * liable for any damages arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to no restriction.
 *
 * Technical remarks and questions can be addressed to:
 * <pornin@bolet.org>
 * -----------------------------------------------------------------------
 */

/*
 * This commnad-line tool generates a new set of delegation parameters,
 * for a provided work factor. The modulus is extracted from an existing
 * encoded modulus, private key or set of delegation parameters.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>

#include "makwa.h"

static void
fail(char *fmt,...)
{
	va_list ap;

	va_start(ap, fmt);
	fprintf(stderr, "FAIL: ");
	vfprintf(stderr, fmt, ap);
	fprintf(stderr, "\n");
	va_end(ap);
	abort();
}

#define CF(x)   do { \
		int error_code = (x); \
		if (error_code < 0) { \
			fail("line %ld: %s returned %d\n", \
				(long)__LINE__, #x, error_code); \
			abort(); \
		} \
	} while (0)

static void *
xmalloc(size_t len)
{
	void *x;

	if (len == 0) {
		return NULL;
	}
	x = malloc(len);
	if (x == NULL) {
		fail("memory allocation failure: %lu bytes\n",
			(unsigned long)len);
	}
	return x;
}

static void
xfree(void *x)
{
	if (x != NULL) {
		free(x);
	}
}

static void
usage(void)
{
	fprintf(stderr,
"usage: deleggen infile workfactor outfile\n");
	exit(EXIT_FAILURE);
}

static void *
read_file(char *name, size_t *data_len)
{
	FILE *f;
	unsigned char *buf;
	size_t ptr, len;

	f = fopen(name, "rb");
	if (f == NULL) {
		fail("could not open file '%s' for reading", name);
	}
	buf = NULL;
	ptr = 0;
	len = 0;
	for (;;) {
		unsigned char tmp[8192];
		size_t rlen, nptr;

		rlen = fread(tmp, 1, sizeof tmp, f);
		if (rlen == 0) {
			break;
		}
		nptr = ptr + rlen;
		if (nptr > len) {
			unsigned char *nbuf;
			size_t nlen;

			nlen = len << 1;
			if (nlen < nptr) {
				nlen = nptr;
			}
			nbuf = xmalloc(nlen);
			memcpy(nbuf, buf, ptr);
			xfree(buf);
			buf = nbuf;
			len = nlen;
		}
		memcpy(buf + ptr, tmp, rlen);
		ptr = nptr;
	}
	*data_len = ptr;
	return buf;
}

static void
write_file(char *name, const void *data, size_t data_len)
{
	FILE *f;

	f = fopen(name, "wb");
	if (f == NULL) {
		fail("could not open file '%s' for writing", name);
	}
	if (fwrite(data, 1, data_len, f) != data_len) {
		fail("short write in file '%s': %lu bytes",
			name, (unsigned long)data_len);
	}
	fclose(f);
}

int
main(int argc, char *argv[])
{
	char *infile, *swf, *outfile;
	long work_factor;
	char *cptr;
	unsigned char *param, *dp;
	size_t param_len, dp_len;

	if (argc != 4) {
		usage();
	}
	infile = argv[1];
	swf = argv[2];
	outfile = argv[3];
	work_factor = strtol(swf, &cptr, 0);
	if (*swf == 0 || *cptr != 0 || work_factor < 0) {
		usage();
	}
	param = read_file(infile, &param_len);
	CF(makwa_delegation_generate(param, param_len,
		work_factor, NULL, &dp_len));
	dp = xmalloc(dp_len);
	CF(makwa_delegation_generate(param, param_len,
		work_factor, dp, &dp_len));
	write_file(outfile, dp, dp_len);
	xfree(param);
	xfree(dp);
	return 0;
}

File: Makwa/java/src/makwa/tools/DelegGen.java
/*
 * -----------------------------------------------------------------------
 * (c) Thomas Pornin 2014. This software is provided 'as-is', without
 * any express or implied warranty. In no event will the author be held
 * liable for any damages arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to no restriction.
 *
 * Technical remarks and questions can be addressed to:
 * <pornin@bolet.org>
 * -----------------------------------------------------------------------
 */

package makwa.tools;

import java.io.ByteArrayOutputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

import makwa.MakwaDelegation;

/**
 * <p>This command-line tools generates sets of parameters for Makwa
 * delegation. Usage:</p>
 * <pre>
 *    DelegGen inParam workFactor outFile
 * </pre>
 *
 * <p>The {@code inParam} parameter is the name of a file containing an
 * encoded Makwa modulus, or an encoded Makwa private key. Processing
 * is (much) faster if a private key is used; but the obtained set is
 * equally valid otherwise.</p>
 *
 * <p>The {@code workFactor} parameter is the work factor for which the
 * set of parameters is created. Each set of parameters is specific to a
 * single work factor.</p>
 *
 * <p>The resulting set of parameters is finally encoded into the
 * file whose name is provided as {@code outFile}.</p>
 *
 * @version   $Revision$
 * @author    Thomas Pornin
 */

public class DelegGen {

	public static void main(String[] args)
		throws IOException
	{
		if (args.length != 3) {
			usage();
		}
		byte[] mparam = readAllBytes(args[0]);
		int workFactor = Integer.parseInt(args[1]);
		MakwaDelegation md = MakwaDelegation.generate(
			mparam, workFactor);
		FileOutputStream out = new FileOutputStream(args[2]);
		try {
			out.write(md.export());
		} finally {
			out.close();
		}
	}

	private static void usage()
	{
		System.err.println(
"usage: DelegGen inParam workFactor outFile");
		System.exit(1);
	}

	private static byte[] readAllBytes(String name)
		throws IOException
	{
		ByteArrayOutputStream baos = new ByteArrayOutputStream();
		byte[] buf = new byte[8192];
		FileInputStream in = new FileInputStream(name);
		try {
			for (;;) {
				int len = in.read(buf);
				if (len < 0) {
					return baos.toByteArray();
				}
				baos.write(buf, 0, len);
			}
		} finally {
			in.close();
		}
	}
}
